gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/MultiLongWatermark.java

```diff
@@ -0,0 +1,159 @@
+/* (c) 2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.source.extractor.extract.kafka;
+
+import java.math.RoundingMode;
+import java.util.List;
+
+import com.beust.jcommander.internal.Lists;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ComparisonChain;
+import com.google.common.math.LongMath;
+import com.google.common.primitives.Shorts;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+
+import gobblin.source.extractor.Watermark;
+
+
+/**
+ * A {@link gobblin.source.extractor.Watermark} that holds multiple long values.
+ *
+ * @author ziliu
+ */
+public class MultiLongWatermark implements Watermark {
+
+  private static final Gson GSON = new Gson();
+
+  private final List<Long> values;
+
+  /**
+   * Copy constructor.
+   */
+  public MultiLongWatermark(MultiLongWatermark other) {
+    this.values = Lists.newArrayList(other.values);
+  }
+
+  public MultiLongWatermark() {
+    this.values = Lists.newArrayList();
+  }
+
+  /**
+   * Increment the idx'th value. idx must be between 0 and size()-1.
+   */
+  public void increment(int idx) {
+    Preconditions.checkElementIndex(idx, this.values.size());
+    Preconditions.checkArgument(this.values.get(idx) < Long.MAX_VALUE);
+    this.values.set(idx, this.values.get(idx) + 1);
+  }
+
+  /**
+   * A MultiLongWatermark w1 is less than w2, if there exists i, such that
+   * the 0th to (i-1)th values of w1 are equivalent as those of w2, and the ith
+   * value of w1 is less than the that of w2.
+   *
+   * @param wm must be an instance of MultiLongWatermark, and this.size() must
+   * equal wm.size().
+   */
+  @Override
+  public int compareTo(Watermark wm) {
```

Comment:
Yes, watermarks need to be comparable in some way, but there's no reason they need to have a total order as is implied by the comparable interface. In the multilong watermark, for example, the watermarks (5,5,5) and (5,6,4) should not be considered sortable. 

Instead, I suggest that watermarks should implement a less restrictive interface that merely requires a "isLargerThan" method, then (5,5,5).isLargerThan(5,5,4) is true, but (5,5,5).isLargerThan(5,6,4) is false, and also (5,6,4).isLargerThan(5,5,5) is false.

The requirement would be that isLargerThan is transitive, and that every time the watermark is incremented it is true that low_watermark < old_watermark < new_watermark < high_watermark, otherwise throw an IllegalStateException. This offers all the requirements explained above, without implying or requiring total order guarantees.